modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -2308,38 +2355,31 @@ private CompletableFuture<BinaryRow> resolveRoReadResult(
      * @param readResult Read result to resolve.
      * @param txId Nullable transaction id, should be provided if resolution is performed within the context of RW transaction.
      * @param timestamp Timestamp is used in RO transaction only.
-     * @param lastCommitted Action to get the latest committed row, it is used in RO transaction only.
+     * @param lastCommitted Action to get the latest committed row.
      * @return Future to resolved binary row.
      */
     private CompletableFuture<BinaryRow> resolveReadResult(
             ReadResult readResult,
             @Nullable UUID txId,
             @Nullable HybridTimestamp timestamp,
-            @Nullable Supplier<BinaryRow> lastCommitted
+            Supplier<BinaryRow> lastCommitted
     ) {
         if (readResult == null) {
             return completedFuture(null);
+        } else if (!readResult.isWriteIntent()) {
+            return completedFuture(readResult.binaryRow());
         } else {
+            // RW resolution.
             if (txId != null) {
```

Comment:
Do we assume that RO transaction has null txId?

Reply:
Yes, it is an old comment

Reply:
Is that actually true?

Reply:
Of couse.

Reply:
According to the TxManagerImpl.begin() it's not true
```
    public InternalTransaction begin(boolean readOnly, @Nullable HybridTimestamp observableTimestamp) {
        assert readOnly || observableTimestamp == null : "Observable timestamp is applicable just for read-only transactions.";

        HybridTimestamp beginTimestamp = clock.now();
        UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp);
        changeState(txId, null, PENDING);

        if (!readOnly) {
            return new ReadWriteTransactionImpl(this, txId);
        }
```

Reply:
It is a different place, and we never pass the transaction id in resolution for RO transactions.

Reply:
It's very unreliable. Do we have any other way to distinguish RW from RO here?

Reply:
Here we can compare the read timestamp with null. The null is only possible for RW transactions.